### ISSUE TITLE
Prometheus exporter produces non cumulative histogram buckets

### ISSUE THREAD

**Issue by rozaliev** [census-instrumentation/opencensus-go#649]
I did some testing with opencensus-go and prometheus library, it seems that opencensus produces non cumulative buckets. 

How to reproduce:
```
exporter, err := prometheus.NewExporter(prometheus.Options{})
if err != nil {
   log.Fatal(err)
}
view.RegisterExporter(exporter)
if err := view.Subscribe(ochttp.DefaultServerViews...); err != nil {
   log.Fatal(err)
}
```

Expected result 
```
opencensus_opencensus_io_http_server_latency_bucket{le="0"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="1"} 1
opencensus_opencensus_io_http_server_latency_bucket{le="2"} 1
opencensus_opencensus_io_http_server_latency_bucket{le="3"} 1
opencensus_opencensus_io_http_server_latency_bucket{le="4"} 1
opencensus_opencensus_io_http_server_latency_bucket{le="5"} 1
opencensus_opencensus_io_http_server_latency_bucket{le="6"} 1
opencensus_opencensus_io_http_server_latency_bucket{le="8"} 1
opencensus_opencensus_io_http_server_latency_bucket{le="10"} 1
opencensus_opencensus_io_http_server_latency_bucket{le="13"} 939126
opencensus_opencensus_io_http_server_latency_bucket{le="16"} 974422
opencensus_opencensus_io_http_server_latency_bucket{le="20"} 974609
opencensus_opencensus_io_http_server_latency_bucket{le="25"} 974609
opencensus_opencensus_io_http_server_latency_bucket{le="30"} 974609
opencensus_opencensus_io_http_server_latency_bucket{le="40"} 974609
opencensus_opencensus_io_http_server_latency_bucket{le="50"} 974609
opencensus_opencensus_io_http_server_latency_bucket{le="65"} 974609
opencensus_opencensus_io_http_server_latency_bucket{le="80"} 974609
opencensus_opencensus_io_http_server_latency_bucket{le="100"} 974609
opencensus_opencensus_io_http_server_latency_bucket{le="130"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="160"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="200"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="250"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="300"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="400"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="500"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="650"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="800"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="1000"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="2000"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="5000"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="10000"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="20000"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="50000"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="100000"} 974611
opencensus_opencensus_io_http_server_latency_bucket{le="+Inf"} 974611
opencensus_opencensus_io_http_server_latency_sum 1.0789131900678957e+07
opencensus_opencensus_io_http_server_latency_count 974611
```
But instead got this
```
opencensus_opencensus_io_http_server_latency_bucket{le="0"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="1"} 1
opencensus_opencensus_io_http_server_latency_bucket{le="2"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="3"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="4"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="5"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="6"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="8"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="10"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="13"} 947983
opencensus_opencensus_io_http_server_latency_bucket{le="16"} 26577
opencensus_opencensus_io_http_server_latency_bucket{le="20"} 48
opencensus_opencensus_io_http_server_latency_bucket{le="25"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="30"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="40"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="50"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="65"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="80"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="100"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="130"} 2
opencensus_opencensus_io_http_server_latency_bucket{le="160"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="200"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="250"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="300"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="400"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="500"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="650"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="800"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="1000"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="2000"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="5000"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="10000"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="20000"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="50000"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="100000"} 0
opencensus_opencensus_io_http_server_latency_bucket{le="+Inf"} 974611
opencensus_opencensus_io_http_server_latency_sum 1.0732690369393207e+07
opencensus_opencensus_io_http_server_latency_count 974611
```

**Comment by rakyll** [comment]
@odeke-em it'd be nice if you can take a look. Otherwise, feel free to ping someone else.

**Comment by odeke-em** [comment]
Thank you for the tag @rakyll, I am just seeing this after sitting down at my computer to get working.

**Comment by odeke-em** [comment]
tl;dr: the bug is that NewConstHistogram requires us to have cumulatively added the values by ourselves but we just added discrete values. We can switch back to NewHistogram and then `.Observe(float64)` which will handle auto-cumulation. This is a regression from https://github.com/census-instrumentation/opencensus-go/pull/191 and the bug is in 
https://github.com/census-instrumentation/opencensus-go/blob/8069fe56546c2cdcfbeb5290f5c6130590927308/exporter/prometheus/prometheus.go#L236-L238
Let's use code that automatically handles that for us instead of hardcoding this cumulation and perhaps missing something else in the future.

### Deep dive
The docs at https://godoc.org/github.com/prometheus/client_golang/prometheus#NewConstHistogram say
```buckets is a map of upper bounds to cumulative counts, excluding the +Inf bucket.```
but we are just inserting individual/discrete counts without adding the previous count
https://github.com/census-instrumentation/opencensus-go/blob/8069fe56546c2cdcfbeb5290f5c6130590927308/exporter/prometheus/prometheus.go#L235-L238

Interestingly we initially had the correct cumulative code in https://github.com/orijtech/opencensus-go/blob/5d6b4e33e55670ed4a16201446407f103c6587b5/exporter/stats/prometheus/prometheus.go#L241-L259
<img width="644" alt="screen shot 2018-03-30 at 2 02 30 am" src="https://user-images.githubusercontent.com/4898263/38130005-908e45ec-33be-11e8-84de-1714a8f46707.png">
but we deemed that API to be too low-level.

A remedy would be to first sort the buckets in ascending mapping to their index in data.CountPerBucket, then add them to the points cumulatively to the previous. However, that's very specialized and seems like duck tape plus already rules out being high level anymore. Let's revert to the old code and set Buckets using `NewHistogram` as that's doesn't leak information about the APIs

**Comment by odeke-em** [comment]
I've mailed out https://github.com/census-instrumentation/opencensus-go/pull/661, which uses the latter solution using the already existent code. I kept the current, because there is perhaps a reason why we chose that API but we don't have a documented reason why so I'd prefer to err on the side of caution here. Please take a look.

**Comment by odeke-em** [comment]
Thank you very much @rozaliev for reporting this issue and for using OpenCensus!

We've fixed this issue with PR https://github.com/census-instrumentation/opencensus-go/pull/661, please get the latest from master and the problem should no longer exist.

**Comment by rozaliev** [comment]
I've just tested it, seems to be working great. Thanks for the fix!
